Remove support for delayed inline scripts

joemcgill opened the pull request:
This removes support for delayed inline scripts when used in conjunction with an `async` or `defer` loading strategy. Unit tests are updated to reflect this change in behavior.

Fixes #69.

joemcgill committed:
Update unit tests for removal of inline script support

joemcgill committed:
Remove support for delayed inline scripts

joemcgill commented:
I think I've got most of this updated, but am still seeing unit test failures with the following test: `Tests_Dependencies_Scripts::test_various_strategy_dependency_chains with data set "nested-aliases"`

The nested aliases are being printed with a `defer` attribute, but I don't think they should be, since a dependent includes an `after` inline script. Going to look into this tomorrow.

Also, while reviewing the logic for this, I noticed that we're duplicating a lot of the same checks in `WP_Scripts::has_only_delayed_dependents` and `WP_Scripts::get_eligible_loading_strategy` for determining if the handle can be delayed. It would be nice to try and consolidate this into a single function that are used in both places, if possible.

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -1148,6 +989,11 @@ private function get_eligible_loading_strategy( $handle ) {
+		}
+
 		// Handle async strategy scenario.
 		if ( 'async' === $intended_strategy && $this->has_only_delayed_dependents( $handle, true ) ) {
```
I believe we also have to update the `has_only_delayed_dependents()` method to return `false` if _any_ dependent has an inline "after" script.

Basically for every dependent handle you also have to call the above `has_inline_script( $handle, 'after' )`.

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -75,15 +75,6 @@ class WP_Scripts extends WP_Dependencies {
-	 * @see WP_Scripts::print_delayed_inline_script_loader()
-	 * @var bool
-	 */
-	public $printed_delayed_inline_script_loader = false;
```
Not directly related to this line, but this PR should also remove the `delayed-inline-script-loader.js` file and any references to it.

joemcgill replied:
Handled in [c6fc771](https://github.com/10up/wordpress-develop/pull/81/commits/c6fc771029aaebf3ea22a8e1bef84f1d0abf7d50)

felixarntz reviewed (changes requested):
@joemcgill This looks solid so far, though I think two things are missing (see below).

10upsimon committed:
Update has_only_delayed_dependents to return false if inline after scripts associated with handle or its dependents.

joemcgill committed:
Remove delayed-inline-script-loader.js and references

joemcgill committed:
Consolidate logic for determining eligible loading strategies

This adds a new private method, `WP_Scripts::filter_eligible_strategies` that is responsible for recursively eliminating possible loading strategies. This removes duplication of logic that previously was in place for checking the eligible strategy of a handle with checking for whether dependents could be deferred.

This also removes `WP_Scripts::has_only_deferrable_dependents`, which is no longer needed.

Unit tests for `has_only_deferrable_dependents` have been updated to test `filter_eligible_strategies` instead.

joemcgill commented:
While reviewing the changes we were making to remove support for delayed inline scripts, I noticed that we were often duplicating the same logic for checking a specific handle that we were using to check whether dependents could be deferred.

I've refactored the logic that determines the eligible loading strategy in [a4eeb98](https://github.com/10up/wordpress-develop/pull/81/commits/a4eeb985fbecbefa5a7744f3e4b1bc9e859048fe). This simplifies things quite a bit by recursively using a process of elimination to filter out eligible strategies for a handle and consolidates all the logic to a new private method, `WP_Scripts::filter_eligible_strategies` which replaces `WP_Scripts::has_only_delayed_dependents`.

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
-	private function has_inline_script( $handle ) {
+	private function has_inline_script( $handle, $position = null ) {
+		if ( $position && in_array( $position, array( 'before', 'after' ), true ) ) {
+			return $this->get_data( $handle, $position );
```
Since per the docs this should return a boolean (which also makes sense given the function name), we have to cast it here to return the correct type.
```suggestion
			return (bool) $this->get_data( $handle, $position );
```

joemcgill replied:
Fixed in [12969ea](https://github.com/10up/wordpress-develop/pull/81/commits/12969ead675c93e2f1aa83bddf038e5f20ab96f0)

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
+		$dependents = $this->get_dependents( $handle );
+
+		// Recursively filter eligible strategies for dependents.
+		foreach ( $dependents as $dependent ) {
+			// Bail early once we know the eligible strategy is blocking.
+			if ( empty( $eligible ) ) {
+				return array();
+			}
+
+			$eligible = $this->filter_eligible_strategies( $dependent, $eligible, $checked );
 		}
```
I'm not completely sure this logic is correct. The way it is written it always returns the eligible strategies for the last dependent, unless any of the dependents returned an empty array.

This may not be a common problem in practice, but I think it is possible for one dependent to e.g. return `array( 'defer' )` and another dependent later to return `array( 'async' )` or `array( 'async', 'defer' )`. Shouldn't the return value then be the stricter one, i.e. `array( 'defer' )`?

joemcgill replied:
We’re passing the array of eligible strategies to this function, which then gets filtered by the characteristics of the dependent, so the filtering is cumulative.

felixarntz replied:
Thanks for clarifying @joemcgill, that makes sense. Just a recommendation then from my end would be to make `$eligible` the first parameter of that method, as that makes it more intuitive that that value is what's being filtered in the method.

joemcgill replied:
The only required param in this method is the first one, `$handle`, so I think it's best to leave that in the first position.

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
 		}
 
-		// Handle async strategy scenario.
-		if ( 'async' === $intended_strategy && $this->has_only_delayed_dependents( $handle, true ) ) {
-			return 'async';
```
This condition I don't see in the modified logic you implemented. Can you clarify why it was removed? It feels risky to me to make such broad changes so shortly before intended commit and include logic changes like this.

joemcgill replied:
I believe this condition is unnecessary after the refactor. Rather than checking for whether all dependents of an async script are also async, we start by assuming async is possible and eliminate that possibility during the process of filtering eligible strategies.

felixarntz replied:
Now that I understand how the filtering bit in the `foreach` loop works, that makes sense 👍

felixarntz commented on src/wp-includes/class-wp-scripts.php:
```diff
@@ -1058,119 +899,106 @@ private function is_delayed_strategy( $strategy ) {
-		 */
-		if ( empty( $intended_strategy ) ) {
-			return '';
+		// For non-alias handles, an empty intended strategy filters all strategies.
```
What does that mean? I'm not sure I understand. Does returning an empty array here mean that the only eligible strategy is "blocking"? Or does it mean any strategy is eligible? If the latter, I think it would be more intuitive here to just return the array that includes `defer` and `async`, or not?

joemcgill replied:
Right. An empty array means that the script should be blocking. However, an alias (I.e., a handle with no `src`) will not have a strategy applied, so we skip over the check in that case.

felixarntz replied:
Thanks @joemcgill, makes more sense now!

felixarntz reviewed (commented):
@joemcgill I only have one small nit-pick regarding the `has_inline_script()` method, but regarding your additional changes, I have a few more concerns. I am unsure this is a good time to implement a notable refactoring like this, since the logic is quite complex and as far as I can tell you didn't just refactor, but also changed the underlying logic in a few places. I left a few more specific comments below.

Unless there's a good reason for the logic changes, it would be safer to focus only one purely the refactoring, since we're very close to an intended commit. Maybe I'm misunderstanding some bits, but it overall strikes me as a risky change to make this late.

10upsimon commented:
@joemcgill I've given this a review and a test locally, and all seems to check out. I'm happy to mark this approved on my end, but would appreciate @felixarntz eyes on it too.

felixarntz reviewed (approved):
@joemcgill After your clarifications, this LGTM!

Please review my outstanding feedback in https://github.com/10up/wordpress-develop/pull/81#pullrequestreview-1492030326 for consideration, but those are more nit-picks than blockers.

joemcgill committed:
Cast return values for has_inline_script to bool

joemcgill committed:
Add additional unit test cases
